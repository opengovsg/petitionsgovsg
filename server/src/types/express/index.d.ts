declare namespace Express {
  interface User {
    id: string
    type: string
    fullname: string
  }
}
