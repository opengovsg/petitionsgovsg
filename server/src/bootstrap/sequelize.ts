import { Sequelize } from 'sequelize'

import {
  defineSignature,
  definePost,
  defineAddressee,
  defineSubscription,
} from '@/models'
import { dbConfig } from './config/database'

export const sequelize = new Sequelize({ ...dbConfig, logging: false })

export const { Addressee } = defineAddressee(sequelize)
export const { Post } = definePost(sequelize, { Addressee })
export const Signature = defineSignature(sequelize, { Post })
export const Subscription = defineSubscription(sequelize, { Post })

export default sequelize
