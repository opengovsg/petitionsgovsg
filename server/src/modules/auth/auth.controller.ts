import { StatusCodes } from 'http-status-codes'
import { isEmpty } from 'lodash'
import passport from 'passport'
import {
  callbackRedirectUnauthorisedURL,
  callbackRedirectURL,
} from '../../bootstrap/config/auth'
import { Message } from 'src/types/message-type'
import { ErrorDto, LoadPublicUserDto, UserAuthType } from '~shared/types/api'
import { createLogger } from '../../bootstrap/logging'
import { UserService } from '../../modules/user/user.service'
import { ControllerHandler } from '../../types/response-handler'

const logger = createLogger(module)

export class AuthController {
  private userService: Public<UserService>

  constructor({ userService }: { userService: Public<UserService> }) {
    this.userService = userService
  }

  /**
   * Fetch logged in user details after being authenticated.
   * @returns 200 with user details
   * @returns 500 if user id not found
   * @returns 500 if database error
   */
  loadUser: ControllerHandler<unknown, LoadPublicUserDto | ErrorDto> = async (
    req,
    res,
  ) => {
    const id = req.user?.id
    const type = req.user?.type
    if (!id) {
      logger.error({
        message: 'User not found after being authenticated',
        meta: {
          function: 'loadUser',
          userId: req.user?.id,
        },
      })
      return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(null)
    }

    if (type === UserAuthType.Public) {
      try {
        const user = await this.userService.loadUser(id)
        if (user) {
          return res.status(StatusCodes.OK).json({ id: user.id })
        } else {
          logger.error({
            message: 'User not found',
            meta: {
              function: 'loadUser',
              userId: req.user?.id,
            },
          })
          return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(null)
        }
      } catch (error) {
        logger.error({
          message: 'Database Error while loading public user',
          meta: {
            function: 'loadUser',
            userId: req.user?.id,
          },
          error,
        })
        return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(null)
      }
    }

    if (type === UserAuthType.Public) {
      return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(null)
    }
  }

  /**
   * Logout
   * @returns 200 if logged out
   */
  handleLogout: ControllerHandler = (req, res) => {
    if (!req.session || isEmpty(req.session)) {
      logger.error({
        message: 'Attempted to sign out without a session',
        meta: {
          function: 'handleLogout',
        },
      })
      return res.sendStatus(StatusCodes.BAD_REQUEST)
    }

    req.session.destroy((error) => {
      if (error) {
        logger.error({
          message: 'Failed to destroy session',
          meta: {
            action: 'handleLogout',
            function: 'handleLogout',
          },
          error,
        })
        return res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Sign out failed' })
      }

      // No error.
      res.clearCookie('connect.sid')
      return res.status(StatusCodes.OK).json({ message: 'Sign out successful' })
    })
  }

  handleSgidLogin: ControllerHandler<
    undefined,
    undefined,
    undefined,
    { redirect: string }
  > = async (req, res, next) => {
    const { redirect } = req.query
    // store redirect to post in state
    if (redirect) {
      passport.authenticate('sgid-with-name', { state: redirect })(
        req,
        res,
        next,
      )
    } else {
      passport.authenticate('sgid-with-name')(req, res, next)
    }
  }
  /**
   * Verify otp received by the user
   * @params code
   * @params state
   * @returns 302 to home page if successful login
   * @returns 302 to unauthorised page if error
   * @returns 302 to sgid auth page if no state or code params received
   */
  handleSgidCallback: ControllerHandler<
    undefined,
    undefined,
    undefined,
    { code: string; state: string | undefined }
  > = async (req, res, next) => {
    const { state } = req.query
    passport.authenticate(
      'sgid-with-name',
      {},
      (error, user, info: Message) => {
        if (error) {
          logger.error({
            message: 'Error while authenticating',
            meta: {
              function: 'handleSgidCallback',
            },
            error,
          })
          return res.redirect(callbackRedirectUnauthorisedURL)
        }
        if (!user) {
          logger.warn({
            message: info.message,
            meta: {
              function: 'handleSgidCallback',
            },
          })
          res.redirect(callbackRedirectUnauthorisedURL)
        }
        req.logIn(user, (error) => {
          if (error) {
            logger.error({
              message: 'Error while logging in',
              meta: {
                function: 'handleSgidCallback',
              },
              error,
            })
            return res.redirect(callbackRedirectUnauthorisedURL)
          }
          //
          /**
           * Regenerate session to mitigate session fixation
           * We regenerate the session upon logging in so an
           * anonymous session cookie cannot be used
           */
          const passportSession = req.session.passport
          req.session.regenerate(function (error) {
            if (error) {
              logger.error({
                message: 'Error while regenerating session',
                meta: {
                  function: 'handleSgidCallback',
                },
                error,
              })
              return res.redirect(callbackRedirectUnauthorisedURL)
            }
            //req.session.passport is now undefined
            req.session.passport = passportSession
            req.session.save(function (error) {
              if (error) {
                logger.error({
                  message: 'Error while saving regenerated session',
                  meta: {
                    function: 'handleSgidCallback',
                  },
                  error,
                })
                return res.redirect(callbackRedirectUnauthorisedURL)
              }
              return res.redirect(callbackRedirectURL(state))
            })
          })
        })
      },
    )(req, res, next)
  }
}
