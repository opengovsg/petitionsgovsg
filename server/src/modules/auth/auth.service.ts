import { Post, PostStatus } from '~shared/types/base'
import { ModelDef } from '../../types/sequelize'

export class AuthService {
  private Post: ModelDef<Post>

  constructor({ Post }: { Post: ModelDef<Post> }) {
    this.Post = Post
  }

  /**
   * Check if a user has permission to edit a post
   * @param userId of user
   * @param postId of post
   * @returns true if user has permission to answer post
   */
  hasPermissionToEditPost = async (
    userId: number,
    postId: number,
  ): Promise<boolean> => {
    // TODO:
    return true
  }

  /**
   * Check if a user is able to view the post
   * @param post post to be viewed
   * @param userId id of user
   * @returns true if user can view post
   */
  verifyUserCanViewPost = async (
    post: Post,
    userId?: string,
  ): Promise<void> => {
    // If post is public, anyone can view
    if (post.status === PostStatus.Open) return

    // If private or archived, must be logged in
    if (!userId)
      throw new Error('User must be logged in to access private post')

    // TODO
    // const user = await this.User.findOne({ where: { id: userId } })

    // if (user) {
    //   // If officer, they may have permission to answer
    //   if (await this.hasPermissionToEditPost(user.id, post.id)) return

    //   // If none of the above, they must have created the post
    //   if (user.id === post.userId) return
    //   throw new Error('User does not have permission to access post')

    return
  }
}
