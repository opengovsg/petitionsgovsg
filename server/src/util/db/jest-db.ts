import { Sequelize } from 'sequelize'
import { Creation, ModelDef } from '@/types/sequelize'
import {
  defineSignature,
  definePost,
  defineAddressee,
  defineSubscription,
} from '@/models'
import { mockPosts } from './data/post'
import { mockAddressees } from './data/addressee'
import { mockSignatures } from './data/signature'
import {
  Signature as SignatureModel,
  Addressee as AddresseeModel,
  Post as PostModel,
  Subscription as SubscriptionModel,
} from '~shared/types/base'

export enum ModelName {
  Signature = 'signature',
  Post = 'post',
  User = 'user',
  Addressee = 'addressee',
  Subscription = 'subscription',
}

export type SequelizeWithModels = {
  sequelize: Sequelize
  Addressee: ModelDef<AddresseeModel, Creation<AddresseeModel>>
  Post: ModelDef<PostModel, Creation<PostModel>>
  Signature: ModelDef<SignatureModel, Creation<SignatureModel>>
  Subscription: ModelDef<SubscriptionModel, Creation<SubscriptionModel>>
}

/**
 * Connect to a in-memory database
 */
export const createTestDatabase = async (): Promise<SequelizeWithModels> => {
  const sequelize = new Sequelize('sqlite::memory:', { logging: false })

  const { Addressee } = defineAddressee(sequelize)
  const { Post } = definePost(sequelize, { Addressee })
  const Signature = defineSignature(sequelize, { Post })
  const Subscription = defineSubscription(sequelize, { Post })

  const models = {
    Addressee,
    Post,
    Signature,
    Subscription,
  }

  const db = { ...models, sequelize }
  await resetAndSetupDb(db)
  return db
}

/**
 * Clears and re-populates database with test fixtures
 */
export const resetAndSetupDb = async (
  db: SequelizeWithModels,
): Promise<void> => {
  await resetTestDb(db)
  await populateTestDb(db)
}

/**
 * Clears database
 */
const resetTestDb = async (
  db: SequelizeWithModels,
): Promise<SequelizeWithModels> => {
  // Set force to true to clear DB if tables exist
  await db.sequelize.sync({ force: true })
  return db
}

/**
 * Populates database with mock data
 */
const populateTestDb = async (
  db: SequelizeWithModels,
): Promise<SequelizeWithModels> => {
  await db.Addressee.bulkCreate(mockAddressees)
  await db.Post.bulkCreate(mockPosts)
  await db.Signature.bulkCreate(mockSignatures)
  return db
}

export function getModelDef<T, C = Creation<T>>(
  sequelize: Sequelize,
  modelName: ModelName,
): ModelDef<T, C> {
  return sequelize.models[modelName] as ModelDef<T, C>
}
