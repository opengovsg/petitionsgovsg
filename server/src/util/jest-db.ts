import { Model, ModelCtor, Sequelize } from 'sequelize'
import { Creation, ModelDef } from '../types/sequelize'
import { defineSignature, definePost } from '../models'

export enum ModelName {
  Signature = 'signature',
  Post = 'post',
  User = 'user',
}

/**
 * Connect to a in-memory database
 */
export const createTestDatabase = async (): Promise<Sequelize> => {
  const sequelize = new Sequelize('sqlite::memory:', { logging: false })

  const { Post } = definePost(sequelize)
  const Signature = defineSignature(sequelize, { Post })

  await sequelize.sync()

  return sequelize
}

export function getModel<T extends Model>(
  sequelize: Sequelize,
  modelName: ModelName,
): ModelCtor<T> {
  return sequelize.models[modelName] as ModelCtor<T>
}

export function getModelDef<T, C = Creation<T>>(
  sequelize: Sequelize,
  modelName: ModelName,
): ModelDef<T, C> {
  return sequelize.models[modelName] as ModelDef<T, C>
}
