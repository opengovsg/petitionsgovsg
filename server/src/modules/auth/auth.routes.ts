import express from 'express'
import { AuthController } from './auth.controller'
import { AuthMiddleware } from './auth.middleware'

export const routeAuth = ({
  controller,
  authMiddleware,
}: {
  controller: AuthController
  authMiddleware: AuthMiddleware
}): express.Router => {
  const router = express.Router()

  /**
   * Fetch logged in user details
   * @route   GET /api/auth
   * @returns 200 with user details
   * @returns 401 if user not signed in
   * @returns 500 if database error
   * @access  Private
   */
  router.get('/', authMiddleware.authenticate, controller.loadUser)

  /**
   * Logout
   * @route   POST /api/auth/logout
   * @returns 200 if logged out
   * @access  private
   */
  router.post('/logout', controller.handleLogout)

  /**
   * Sgid Login
   * @route   POST /api/auth/sgid/login
   * @returns 302 with sgid oidc link
   * @access  private
   */
  router.get('/sgid/login', controller.handleSgidLogin)

  /**
   * Sgid Callback
   * @route   GET /api/auth/callback
   * @returns 302 to home page if successful login
   * @returns 302 to unauthorised page if error
   * @returns 302 to sgid auth page if no state or code params received
   * @access  private
   */
  router.get('/callback', controller.handleSgidCallback)

  return router
}
