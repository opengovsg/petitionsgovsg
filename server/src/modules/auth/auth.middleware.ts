import { StatusCodes } from 'http-status-codes'
import { ControllerHandler } from '../../types/response-handler'
import { decodeUserJWT } from '../../util/jwt'
export class AuthMiddleware {
  /**
   * Middleware that only allows authenticated users to pass through to the next
   * handler.
   * @returns next if user exists in session
   * @returns 401 if user does not exist in session
   */
  authenticate: ControllerHandler = (req, res, next) => {
    // Check if user is authenticated
    try {
      const token = req.cookies.jwt
      req.user = decodeUserJWT(req)
    } catch (error) {
      return res
        .status(StatusCodes.UNAUTHORIZED)
        .json({ message: 'User not signed in.' })
    }
    return next()
  }
}
