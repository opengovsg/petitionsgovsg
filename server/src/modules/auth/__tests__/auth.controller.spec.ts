import express from 'express'
import { StatusCodes } from 'http-status-codes'
import supertest from 'supertest'
import { ControllerHandler } from '../../../types/response-handler'
import { AuthController } from '../auth.controller'
import { Sequelize } from 'sequelize'
import { createTestDatabase } from '../../../util/jest-db'
import { UserAuthType } from '~shared/types/api'

jest.mock('../../../util/hash')

// Mock services invoked by authController
const userService = {
  loadUser: jest.fn(),
  loadUserBySgid: jest.fn(),
}

let authController: AuthController

const path = '/auth'
// Set up auth middleware to inject user
const user: Express.User | undefined = { id: 1, type: UserAuthType.Public }
const isAuthenticated = true
const middleware: ControllerHandler = (req, res, next) => {
  req.isAuthenticated = () => isAuthenticated
  req.session = {
    passport: { id: 1 },
    regenerate: (callback) => {
      callback(null)
      return req.session
    },
    save: (callback) => {
      if (callback) callback(null)
      return req.session
    },
    id: '',
    cookie: { originalMaxAge: 240000 },
    destroy: jest.fn(),
    reload: jest.fn(),
    resetMaxAge: jest.fn(),
    touch: jest.fn(),
  }
  req.logIn = (_, callback) => {
    callback()
  }
  ;(req.user = user), next()
}

// Set up supertest
const app = express()
app.use(express.json())
app.use(middleware)
const request = supertest(app)

// Set up sequelize
let db: Sequelize

beforeAll(async () => {
  db = await createTestDatabase()
  authController = new AuthController({
    userService,
  })
  app.get(path, authController.loadUser)
  app.get(path + '/logout', authController.handleLogout)
})

afterEach(() => {
  jest.clearAllMocks()
})

afterAll(async () => {
  await db.close()
})

describe('auth.controller', () => {
  describe('loadUser', () => {
    it('should return 200 when public user id is valid', async () => {
      // Arrange
      const userData = {
        sgid: 'u=35',
        displayname: 'LIM YONG XIANG',
        fullname: 'LIM YONG XIANG',
        email: 'limyongxiang@test.gov.sg',
        active: true,
      }
      userService.loadUser.mockReturnValue(userData)

      // Act
      const response = await request.get(path)

      // Assert
      expect(response.status).toEqual(StatusCodes.OK)
      expect(response.body).toStrictEqual(userData)
    })

    it('should return 500 if database error', async () => {
      // Arrange
      const error = new Error('Database error')
      userService.loadUser.mockImplementation(() => {
        throw error
      })

      // Act
      const response = await request.get(path)

      // Assert
      expect(response.status).toEqual(StatusCodes.INTERNAL_SERVER_ERROR)
      expect(response.body).toStrictEqual(null)
    })
  })
})
