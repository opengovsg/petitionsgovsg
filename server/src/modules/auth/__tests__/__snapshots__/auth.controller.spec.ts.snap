// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthController handleSgidLogin should return 200 on valid login 1`] = `"http://localhost:5156/sgid/v1/oauth/authorize?client_id=petitionsgov&scope=openid%20myinfo.name&response_type=code&redirect_uri=http%3A%2F%2Flocalhost%3A3000&nonce=abc&state=%2Fcreate"`;

exports[`AuthController loadUser should return 200 when jwt is valid 1`] = `
Object {
  "fullname": "John Doe",
  "id": "sgid",
}
`;

exports[`AuthController loadUser should return 500 when jwt is invalid 1`] = `
Object {
  "message": "Server Error",
}
`;

exports[`AuthController loadUserName should return 200 with user fullname 1`] = `
Object {
  "fullname": "John Doe",
}
`;

exports[`AuthController loadUserName should return 500 on bad service 1`] = `
Object {
  "message": "Server Error",
}
`;

exports[`AuthController loadUserName should return 500 when fullname is not found 1`] = `null`;

exports[`AuthController verifyPetitionOwner should return 200 false when user is not post owner 1`] = `false`;

exports[`AuthController verifyPetitionOwner should return 200 true when user is post owner 1`] = `true`;

exports[`AuthController verifyPetitionOwner should return 500 on bad service 1`] = `
Object {
  "message": "Server Error",
}
`;
